lib/internal/bootstrap/node.js

```diff
@@ -426,23 +426,23 @@
       emitAfter
     } = NativeModule.require('internal/async_hooks');
 
-    process._fatalException = function(er) {
+    process._fatalException = function(er, fromPromise) {
```

Comment:
The only point of `fromPromise` is skipping the `uncaughtException` handler – is that correct? If so, can you explain the motivation behind it?

(Also, if that is correct: Could we name it something like `skipUncaughtExceptionListener` or so, to be a bit more descriptive?)

Reply:
That is correct. In this case I just used the original implementation. I personally have no strong opinion about having a `uncaughtException` in this case. In a way it would make sense to have that.

Reply:
I'm +1 on calling `uncaughtException` handlers on unhandled rejections. I think it would be very surprising to not have them called - especially for APMs.

Reply:
I'm +1 to route unhandled rejection into `'uncaughtException'` at this point.

Reply:
I’d very much prefer this not be the case; I’d like to handle unhandled sync exceptions differently than unhandled promise rejections, always.

If it’s routed into uncaughtException, will there be enough info in the arguments for me to distinguish?

Reply:
@ljharb if you want to distinguish it - you can also do so with an `unhandledRejection` handler - this is just about the default behavior.

Reply:
Routing unhandled rejections into `uncaughtException` is going to have the following behavior:

First it triggers a unhandled rejection and when the GC kicks in, it will also trigger a uncaught exception. Without a unhandled exception handler the process is going to exit. Otherwise it is going to continue. If there is no unhandled rejection handler, it will just keep on printing the warnings but it has no further influence besides that.

Reply:
How can i distinguish a synchronous exception from an Promise rejection in an uncaught exception handler?

Reply:
I think that if you have an `unhandledRejection` handler that should trigger whenever the warning does OR GC does (once per promise) and if it's handled from the handler then `uncaughtException` should not be triggered. If no handler exists or the handler throws the exception back - then `uncaughtException` should be called with it.

Reply:
@ljharb that would not be possible with the suggested implementation. What could be done is to wrap the error into another error that reports about being a unhandled rejection.

@benjamingr that would not be the case with the suggested change. It would work as I explained above. So having a `unhandledRejection` handler will not prevent the process to exit. That would completely depend on `uncaughtException`.

Reply:
@BridgeAR I understand and I don't think that's correct semantics. I think it would surprise users - you'd expect to be able to intercept any `unhandledRejection`. It's very common to use it to `throw` and cause the `uncaughtException` today - or suppress termination in certain cases like:

```
process.on('unhandledRejection', err => {
  if (err.isImportant) log(err);
  if (err.isSpecificKind) return; // not an error, a bug with some library I'm using
  throw err; 
});
```

Reply:
> It is possible to intercept the unhandled rejections. But the error on GC is a different story.

Why? 

> By the way: without routing the unhandled rejection to uncaught exception, it would not be possible to stop the process from exiting.

Which is unfortunate, IMO. 

> It is a anti pattern to continue the program in case a exception is thrown. The pattern about ignoring specific errors is not something we should promote out of my perspective. In case a library is handling something wrong, the library should be fixed or not used.

From the user's perspective - if they're using a library that worked in Node 10 but will break with this PR in Node 12 - you just broke their workflow. The fact _we_ don't like doing it doesn't mean we can drop it. Also note there are cases where you _have_ to keep running after an `unhandledRejection` (or `uncaughtException`) - I can dig up the discussion with Uber about it from a while ago if you'd like.